## Week 9 Homework

1. Download the UCI Heart Disease dataset on Kaggle:  
    https://www.kaggle.com/datasets/cherngs/heart-disease-cleveland-uci  (current link)  
    https://www.kaggle.com/ronitf/heart-disease-uci (old removed link)
    

2. Calculate the accuracy of each of the following at estimating the presence of heart disease
    * Logistic Regression
    * Random Forest
    * Support Vector Machine
    * KNeighbors
    * Decision Tree
    * Ada Boost
    * Bagging
    * Gradient Boosting
    * XGBoost

In [3]:
import statsmodels.api as sm
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier

import xgboost as xgb
from xgboost.sklearn import XGBClassifier

import pandas
from sklearn.svm import SVC
import sklearn

from sklearn.model_selection import train_test_split

In [4]:
col_names = [
    "age", "sex", "cp", "trestbps", "chol",
    "fbs", "restecg", "thalach", "exang", "oldpeak",
    "slope", "ca", "thal", "condition"
]

df = pandas.read_csv(r'C:\Users\khalil\ML HW 9\heart.csv')
df.columns=col_names

df.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
292   40    1   3       152   223    0        0      181      0      0.0   
293   39    1   3       118   219    0        0      140      0      1.2   
294   35    1   3       120   198    0        0      130      1      1.6   
295   35    0   3       138   183    0        0      182      0      1.4   
296   35    1   3       126   282    0        2      156      1      0.0   

     slope  ca  thal  condition  
292      0   0     2          1  
293      1   0     2          1  
294      1   0     2          1  
295      0   0     0          0  
296      0   0     2          1

In [12]:
X = df.drop(columns='condition')
y = df['condition']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=555, stratify=y)

In [13]:
%%time
LR  = LogisticRegression(solver='lbfgs', max_iter=10000, random_state=555)
RF  = RandomForestClassifier(n_estimators = 100, random_state=555)
SVM = SVC(random_state=0, probability=True)
KNC = KNeighborsClassifier()
DTC = DecisionTreeClassifier()
ABC = AdaBoostClassifier(n_estimators = 100)
BC  = BaggingClassifier(n_estimators = 100)
GBC = GradientBoostingClassifier(n_estimators = 100)
#clf_XGB = XGBClassifier(n_estimators = 100, objective= 'binary:logistic', seed=555, use_label_encoder=False)
clf_XGB = XGBClassifier(n_estimators = 100, seed=555, use_label_encoder=False, eval_metric='logloss')
clfs = []
print('5-fold cross validation:\n')
for clf, label in zip([LR, RF, KNC, DTC, ABC, BC, GBC, clf_XGB],
                      ['Logistic Regression',
                       'Random Forest',
                       'Support Vector Machine',
                       'KNeighbors',
                       'Decision Tree',
                       'Ada Boost',
                       'Bagging',
                       'Gradient Boosting',
                       'XGBoost']):
    scores = sklearn.model_selection.cross_val_score(clf, X_train, y_train, cv=5, scoring="accuracy")
    print("Train CV Accuracy: %0.3f (+/- %0.3f) [%s]" % (scores.mean(), scores.std(), label))
    md = clf.fit(X_train, y_train)
    clfs.append(md)
    print("Test Accuracy: %0.4f " % (sklearn.metrics.accuracy_score(clf.predict(X_test), y_test)))

C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")


5-fold cross validation:

Train CV Accuracy: 0.829 (+/- 0.038) [Logistic Regression]
Test Accuracy: 0.8533 
Train CV Accuracy: 0.815 (+/- 0.060) [Random Forest]
Test Accuracy: 0.8667 
Train CV Accuracy: 0.639 (+/- 0.076) [Support Vector Machine]
Test Accuracy: 0.7067 
Train CV Accuracy: 0.784 (+/- 0.056) [KNeighbors]
Test Accuracy: 0.7333 
Train CV Accuracy: 0.748 (+/- 0.030) [Decision Tree]
Test Accuracy: 0.8267 
Train CV Accuracy: 0.811 (+/- 0.052) [Ada Boost]
Test Accuracy: 0.8400 
Train CV Accuracy: 0.806 (+/- 0.035) [Bagging]
Test Accuracy: 0.8400 


C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")


Train CV Accuracy: 0.806 (+/- 0.047) [Gradient Boosting]
Test Accuracy: 0.8000 
Wall time: 6.61 s


3. Which model performed best in #2 ?

Random Forest have highest Accuracy (86%) Among All other algorithms.

4. Find a preprocessing technique we have used in class to improve your best model from above.

In [14]:
from sklearn.preprocessing import LabelEncoder
df1=df.copy() 
df1=df1.apply(LabelEncoder().fit_transform)
df1.tail()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
292    6    1   3        35    54    0        0       79      0        0   
293    5    1   3        13    50    0        0       39      0       12   
294    2    1   3        14    30    0        0       30      1       16   
295    2    0   3        27    19    0        0       80      0       14   
296    2    1   3        19   108    0        2       55      1        0   

     slope  ca  thal  condition  
292      0   0     2          1  
293      1   0     2          1  
294      1   0     2          1  
295      0   0     0          0  
296      0   0     2          1

In [16]:
x1 = df1.drop(columns='condition')
y1 = df1['condition']
x1_train, x1_test, y1_train, y1_test = train_test_split(x1, y1, test_size=0.25, random_state=555, stratify=y1)

In [17]:
%%time
RF  = RandomForestClassifier(n_estimators = 100, random_state=555)
clfs = []
print('5-fold cross validation:\n')
for clf, label in zip([RF],
                      [
                       'Random Forest'
                       ]):
    scores = sklearn.model_selection.cross_val_score(clf, x1_train, y1_train, cv=5, scoring="accuracy")
    print("Train CV Accuracy: %0.3f (+/- %0.3f) [%s]" % (scores.mean(), scores.std(), label))
    md = clf.fit(x1_train, y1_train)
    clfs.append(md)
    print("Test Accuracy: %0.4f " % (sklearn.metrics.accuracy_score(clf.predict(x1_test), y1_test)))

5-fold cross validation:

Train CV Accuracy: 0.838 (+/- 0.032) [Random Forest]
Test Accuracy: 0.8800 
Wall time: 675 ms


Yeah By the use of Pre-processing Technique which I select Label Encoder The Accuracy of my best model which is Random Forest increases by 2% from 86 to 88%

5. Use the ML-Ensemble library to combine your best model and your second best model into a single layer.  
    http://ml-ensemble.com/

In [ ]:
My 1st Best Model is Random Forest and 2nd is Logistic Regression

In [10]:
from mlens.ensemble import SuperLearner
from mlens.model_selection import Evaluator
from mlens.metrics import make_scorer

from sklearn.metrics import accuracy_score

[MLENS] backend: threading


In [20]:
ensemble = SuperLearner(scorer=accuracy_score, random_state=555, verbose=2)

In [21]:
ensemble.add([RF, LR])

SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...42ACB80>)],
   n_jobs=-1, name='group-0', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000001F3742ACB80>,
       shuffle=False, verbose=2)

In [24]:
ensemble.add_meta(GradientBoostingClassifier())

SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...42ACB80>)],
   n_jobs=-1, name='group-1', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000001F3742ACB80>,
       shuffle=False, verbose=2)

In [25]:
ensemble.fit(X_train, y_train)


Fitting 2 layers
Processing layer-1             done | 00:00:01
Processing layer-2             done | 00:00:00
Fit complete                        | 00:00:01


SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...42ACB80>)],
   n_jobs=-1, name='group-1', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000001F3742ACB80>,
       shuffle=False, verbose=2)

In [26]:
print ("Accuracy - Train : ", sklearn.metrics.accuracy_score(ensemble.predict(X_train), y_train))
print ("Accuracy - Test : ", sklearn.metrics.accuracy_score(ensemble.predict(X_test), y_test))


Predicting 2 layers
Processing layer-1             done | 00:00:00
Processing layer-2             done | 00:00:00
Predict complete                    | 00:00:00
Accuracy - Train :  1.0

Predicting 2 layers
Processing layer-1             done | 00:00:00
Processing layer-2             done | 00:00:00
Predict complete                    | 00:00:00
Accuracy - Test :  0.8666666666666667


In [27]:
print("Fit data:\n%r" % ensemble.data)

Fit data:
                                   score-m  score-s  ft-m  ft-s  pt-m  pt-s
layer-1  logisticregression           0.81     0.01  0.58  0.08  0.02  0.02
layer-1  randomforestclassifier       0.81     0.01  0.58  0.07  0.02  0.00



6. Create a custom model using any technique you prefer.  The (two) students who find the best two models in the class will receive +2% on their final exam score.  Any python ML model using techniques we have studied in class are acceptable.

In [ ]:
#Pre-Processing
from sklearn.preprocessing import LabelEncoder
df2=df.copy() 
df2=df2.apply(LabelEncoder().fit_transform)
df2.tail()

In [5]:
df2=df.copy()
x2 = df2.drop(columns='condition')
y2 = df2['condition']
x2_train, x2_test, y2_train, y2_test = train_test_split(x2, y2, test_size=0.20, random_state=555, stratify=y2)

In [6]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
# Fit only on X_train
scaler.fit(x2_train)

# Scale both X_train and X_test
x2_train = scaler.transform(x2_train)
x2_test = scaler.transform(x2_test)

In [ ]:
x2 = df2.drop(columns='condition')
y2 = df2['condition']
x2_train, x2_test, y2_train, y2_test = train_test_split(x2, y2, test_size=0.20, random_state=555, stratify=y2)

In [8]:
RF  = RandomForestClassifier(n_estimators = 100, random_state=555)
clf_XGB = XGBClassifier(n_estimators = 100, seed=555, use_label_encoder=False, eval_metric='logloss')
clfs = []
print('5-fold cross validation:\n')
for clf, label in zip([ RF, clf_XGB],
                      [
                       'Random Forest',
                       'XGBoost']):
    scores = sklearn.model_selection.cross_val_score(clf, x2_train, y2_train, cv=5, scoring="accuracy")
    print("Train CV Accuracy: %0.3f (+/- %0.3f) [%s]" % (scores.mean(), scores.std(), label))
    md = clf.fit(x2_train, y2_train)
    clfs.append(md)
    print("Test Accuracy: %0.4f " % (sklearn.metrics.accuracy_score(clf.predict(x2_test), y2_test)))

5-fold cross validation:

Train CV Accuracy: 0.818 (+/- 0.018) [Random Forest]
Test Accuracy: 0.8167 


C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")
C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")


Train CV Accuracy: 0.802 (+/- 0.029) [XGBoost]
Test Accuracy: 0.7833 


In [11]:
ensemble = SuperLearner(scorer=accuracy_score, random_state=555, verbose=2)

In [13]:
ensemble.add([RF, clf_XGB])

C:\Users\khalil\anaconda3\lib\site-packages\xgboost\sklearn.py:1395: UserWarning: `use_label_encoder` is deprecated in 1.7.0.
  warnings.warn("`use_label_encoder` is deprecated in 1.7.0.")


SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...CECA8B0>)],
   n_jobs=-1, name='group-0', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000002AEECECA8B0>,
       shuffle=False, verbose=2)

In [14]:
ensemble.add_meta(GradientBoostingClassifier())

SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...CECA8B0>)],
   n_jobs=-1, name='group-1', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000002AEECECA8B0>,
       shuffle=False, verbose=2)

In [16]:
ensemble.fit(x2_train, y2_train)


Fitting 2 layers
Processing layer-1             done | 00:00:01
Processing layer-2             done | 00:00:00
Fit complete                        | 00:00:01


SuperLearner(array_check=None, backend=None, folds=2,
       layers=[Layer(backend='threading', dtype=<class 'numpy.float32'>, n_jobs=-1,
   name='layer-1', propagate_features=None, raise_on_exception=True,
   random_state=4782, shuffle=False,
   stack=[Group(backend='threading', dtype=<class 'numpy.float32'>,
   indexer=FoldIndex(X=None, folds=2, raise_on_ex...CECA8B0>)],
   n_jobs=-1, name='group-1', raise_on_exception=True, transformers=[])],
   verbose=1)],
       model_selection=False, n_jobs=None, raise_on_exception=True,
       random_state=555, sample_size=20,
       scorer=<function accuracy_score at 0x000002AEECECA8B0>,
       shuffle=False, verbose=2)

In [18]:
print ("Accuracy - Train : ", sklearn.metrics.accuracy_score(ensemble.predict(x2_train), y2_train))
print ("Accuracy - Test : ", sklearn.metrics.accuracy_score(ensemble.predict(x2_test), y2_test))


Predicting 2 layers
Processing layer-1             done | 00:00:00
Processing layer-2             done | 00:00:00
Predict complete                    | 00:00:00
Accuracy - Train :  1.0

Predicting 2 layers
Processing layer-1             done | 00:00:00
Processing layer-2             done | 00:00:00
Predict complete                    | 00:00:00
Accuracy - Test :  0.8166666666666667


In [19]:
print("Fit data:\n%r" % ensemble.data)

Fit data:
                                   score-m  score-s  ft-m  ft-s  pt-m  pt-s
layer-1  randomforestclassifier       0.84     0.02  0.88  0.05  0.09  0.00
layer-1  xgbclassifier                0.79     0.02  0.27  0.04  0.00  0.00



As I tried different Methods having different percentages of Test-Train sets but the highest Test Accuracy that I got is 79% but 100% Train set Accuracy (But I am in doubt that the model may go into overfitting)